bin/test-cluster.sh

```diff
@@ -2,41 +2,131 @@
 
 DIR=$(dirname $0)
 cmd=$1
+defaultServerPortBase=55180
+defaultStoragePortBase=55280
 
-start() {
+# supported commands: start/stop/restart/clean
+# test-cluster.sh start <- initiate network, create a default cluster
+# test-cluster.sh start <cluster_name> <- starts again already created cluster on the same ports
+# test-cluster.sh start <cluster_name> <base_server_port> <base_storage_port> <- initiate network,
+# adds new cluster of one storage & server node running on provided ports
+# test-cluster.sh stop (cluster_name) <- stop all clusters/(stop cluster with given cluster name)
+# test-cluster.sh restart (cluster_name) <- restart all clusters/(restart cluster with given cluster name)
+# test-cluster.sh clean (cluster_name) <- remove all clusters/(remove cluster with given cluster name)
+
+initNetwork() {
     $DIR/docker/create-network.sh
     $DIR/docker/zookeeper.sh start
-    $DIR/docker/cluster.sh create
-    $DIR/docker/waltz-storage.sh start
-    $DIR/docker/add-storage.sh
-    $DIR/docker/waltz-server.sh start
+}
+
+getContainerPort() {
+    containerName=$1
+    containerExists=$(docker container ls -a --format '{{.Names}}' --filter "name=$containerName" | wc -l)
+    if [ "$containerExists" -gt 0 ]; then
+        containerId=$(docker container ls -a --format '{{.ID}}' --filter "name=$containerName" | head -1)
+        assignedPort=$(docker inspect --format='{{.HostConfig.PortBindings}}' $containerId | grep -o -E '[0-9]+' | head -1)
+        echo "$assignedPort"
+    else
+        echo "docker container $containerName not found"
+        exit 1
+    fi
+}
+
+rerun() {
+    clusterName=$1
+    serverPortBase=$(getContainerPort "${clusterName}_server")
+    storePortBase=$(getContainerPort "${clusterName}_storage")
+    startCluster $clusterName $serverPortBase $storePortBase
+}
+
+startCluster() {
+    clusterName=$1
+    serverPortLowerBound="$2"
+    storagePortLowerBound="$3"
+    $DIR/docker/cluster-config-files.sh "$clusterName" $serverPortLowerBound $storagePortLowerBound
+
+    $DIR/docker/cluster.sh create "$clusterName"
+    $DIR/docker/waltz-storage.sh start "$clusterName" $storagePortLowerBound
+    $DIR/docker/add-storage.sh "$clusterName" $storagePortLowerBound
+    $DIR/docker/waltz-server.sh start "$clusterName" $serverPortLowerBound
+}
+
+stopCluster() {
+    $DIR/docker/waltz-server.sh stop "$1"
```

Comment:
Just had a question regarding stop and clean, in `waltz-server.sh` and also in storage, it is and was not specifically checking the passed-in `$cmd` with "stop" nor use it to do anything, just curious how the stop/clean was handled? What does those scripts do for a stop/clean? Not sure if I'm missing anything.

Reply:
the commands are triggered in container.sh. Basically command `source $DIR/container.sh` in `waltz-server.sh` extends `waltz-server.sh` with `container.sh`.